rusty-plants/src/main.rs

```diff
@@ -23,37 +26,52 @@ struct Cli {
     genome_path: PathBuf,
 }
 
+
 fn main() {
     prisma_client_rust_cli::run();
 
+    let safe_memory_limit = get_safe_memory_limit();
+    println!("Safe memory limit: {} KB", safe_memory_limit);
+
     let read_dir = std::path::Path::new("../data/reads/");
     let files = read_directory_to_string(read_dir).expect("failed to read fragment files");
     let fragments = parse_file(&files).expect("failed to parse fragments");
-    let genome =
-        parse_genome(read_to_string("../data/ref_genome.fna").expect("failed to read genome file"));
+    let genome = parse_genome(read_to_string("../data/ref_genome.fna").expect("failed to read genome file"));
     let transcriptome = Transcriptome::new(&genome);
-    let mut st = suffix_tree::SuffixTree::new();
-    for (i, c) in transcriptome
-        .get_bases()
-        .chars()
-        .map(|x| x as u8)
-        .take(80_000_000)
-        .enumerate()
-    {
-        if i % 1_000_000 == 0 {
-            println!("Added up to transcriptome character {} to suffix tree", i);
-        }
-        st.extend(c);
-    }
 
-    for i in 0..fragments.len() {
-        let matches = align_fragment(&fragments[i], &st, &transcriptome);
-        let mut best: &Match = &matches[0];
-        for m in matches.iter() {
-            if m.errors < best.errors {
-                best = &m;
+    // Calculate max_chars and total_bases
+    let max_chars = (safe_memory_limit * 1024) / (std::mem::size_of::<u8>() as u64) / 2; // Reduce chunk size
+    let total_bases = transcriptome.get_bases().chars().count();
+    let num_chunks = (total_bases as f64 / max_chars as f64).ceil() as usize;
+
+    for chunk in 0..num_chunks {
+        let mut st = SuffixTree::new();
+        let start = chunk * max_chars as usize;
+        let end = ((chunk + 1) * max_chars as usize).min(total_bases);
+
+        println!("Processing chunk from {} to {}", start, end);
+        process_chunk(&transcriptome, &mut st, start, end);
+
+        let sys = sysinfo::System::new_all();
+        println!("Memory usage after processing chunk {}: {} KB", chunk, sys.used_memory());
+
+        println!("Aligning fragments");
+        for i in 0..fragments.len() {
```

Comment:
with it now being multiple suffix trees were now checking this against multiple times. A natural consequence.

Reply:
Sufix tree and sys moved outside for loop.